### DISCUSSION TITLE
Ability to use script tags

### DISCUSSION THREAD

**Discussion by Director-Fusion** [Syslifters/sysreptor#98] [General]
Without being able to run JS via script tags the chartjs feature is severely limited in functionality. Being able to have the ability to allow the use of script tags (optional on/off) OR if worried about security with the product allow the import of server-side scripting to import a more robust chart that can iterate through the arrays. I'd rather be limited by imagination and skill set than by the tool.

**Comment by aronmolnar** [comment]
Hi,
Can you explain your exact use case?

Can you prepare variables in JS and use them in your charts?

**Reply by Director-Fusion** [reply]
Yes exactly, I would be able to run JS in the scirpt tags to iterate through my findings to create different charts to convey a better visual for clients. Right now I am limited to whats exported by the tool. Where I can generate charts based on different aspects of the findings to create better visuals for the clients that need to see that. Yeah for most of us it is just added flash but for clients visuals can be very useful. This unlocks the full potential of chartjs in the reporting software. 

I can't do a for loop via a vuejs tag to iterate through data that will just create a chart for each piece of data. So doing a v-for loop in vue js in the div or chart tag of my chart would do a new chart for each cvss score or what other item i want out of the data available. so 3 findings would create 3 charts showing the cvss score on the chart. 

But by using JS inside a script tag I could generate that chart with all my cvss scores on one chart. Im not hung up on cvss scores but that is an example of what is already being exported. I would do different data based on organizational wants.

**Reply by aronmolnar** [reply]
Currently, it is only possible to run JavaScript inline in your charts. You can therefore use the lodash syntax or JavaScript operations.

Would you provide more detailed information, what kind of chart you are trying to build? Maybe also an example chart (you can do it in Excel) to make clear what the chart should show.  
We might be able to support you in building your chart.

**Reply by Director-Fusion** [reply]
I am building a scatter chart. I am just wanting to iterate through the cvss scores to place in a chart to generate a visual for clients. The chart is done, aside from the iteration issue. I can iterate with the v-for. But that creates a chart for each finding cvss score. I want the cvss scores all on one scatter chart. I guess how would I go about using the datalabels to get finding.title as the label and x,y finding.cvss.score with this lodash feature you mentioned? I gave it a go and it gives and error when doing the function JS placeholder like lodash.forEach(findings, function(finding). 

The current chart is using the v-for iterator in the div tag. How can i get all the cvss scores in one chart instead of one per finding with the lodash utilities mentioned earlier? Thanks for your help with it. This is a very cool tool and just trying to make my reports stand out.
![image](https://github.com/Syslifters/sysreptor/assets/111129863/77f3f852-29ca-4aa3-9430-2d76d40da1c0)

**Reply by aronmolnar** [reply]
What are your axes?
CVSS and what else?

**Reply by Director-Fusion** [reply]
Apologies, CVSS was just an arbitrary bit to explain how i need to iterate through the findings portion of the JSON data. We are using custom fields added to the report. X- Damage, Y - Exploitability(how difficult to exploit). When iterated it will be "finding.damage" or "finding.exploitability" respectively. Just a integer. Exactly as you see in the graph. Apologies for the screenshot, I didn't realize that the axes were cut off from the screenshot of the graph. We are using a gradient via CSS to act as a sort of heat map for the findings. To give the client the idea of what may be more pressing not just so much this could be bad.
![heatmap](https://github.com/Syslifters/sysreptor/assets/111129863/6f2cee3c-dd62-4abd-b0d1-634e04d0ea8a)

**Reply by Director-Fusion** [reply]
SOOOOO. I figured it out using the JS map func. 

` data: findings.map((finding) => ({
                            label: finding.title, 
                            x: finding.damage, 
                            y: finding.exploitability
                            })),`

My issue was stemming from the data field having square brackets in there. Once I got the syntax for lodash and now just the JS map down, i played around with it and removed the square brackets. This populated the data on one chart as desired. Apologies for my lack of understanding of the product before raising the question. Thanks for being helpful none the less. Greatly appreciated.

Data is not from a client just arbitrary ints put in for report designing. Don't take the values to heart.

![working heatmap](https://github.com/Syslifters/sysreptor/assets/111129863/5a7e9907-e36e-49e5-818c-2bb439491fc9)

**Reply by aronmolnar** [reply]
Great - very good to hear. Congratulations.
No problem.

**Comment by aronmolnar** [comment]
This feature is now implemented.
